src/utilities/geometry/Polygon3d.cpp

```diff
@@ -0,0 +1,236 @@
+  }
+}
+
+/// Adds a point to the polygon perimeter
+void Polygon3d::addPoint(const Point3d& point) {
+  m_outerPath.emplace_back(point);
+}
+
+/// Sets the perimeter of the polygon
+void Polygon3d::setOuterPath(const Point3dVector& outerPath) {
+  m_outerPath = outerPath;
+}
+
+Point3dVector Polygon3d::getOuterPath() const {
+  return m_outerPath;
+}
+
+Point3dVectorVector Polygon3d::getInnerPaths() const {
+  return m_innerPaths;
+}
+
+void Polygon3d::addHole(const Point3dVector& hole) {
+  m_innerPaths.emplace_back(hole);
+}
+
+Vector3d Polygon3d::newallVector() {
+  if (OptionalVector3d v = openstudio::getNewallVector(m_outerPath)) {
+    return v.get();
+  }
+
+  return Vector3d();
+}
+
+Vector3d Polygon3d::outwardNormal() {
+  return openstudio::getOutwardNormal(m_outerPath).get();
+}
+
+double Polygon3d::grossArea() {
+
+  if (boost::optional<double> area = openstudio::getArea(m_outerPath)) {
+    return area.get();
+  }
+
+  return 0;
+}
+
+double Polygon3d::netArea() {
+
+  double netArea = grossArea();
+  for (const auto& hole : m_innerPaths) {
+    if (boost::optional<double> area = openstudio::getArea(hole)) {
+      netArea -= area.get();
+    }
+  }
+
+  return netArea;
+}
+
+double Polygon3d::perimeter() {
+
+  double perimeter = 0;
+  for (size_t i = 0; i < m_outerPath.size(); i++) {
+    Point3d p1 = m_outerPath[i];
+    Point3d p2 = m_outerPath[(i + 1) % m_outerPath.size()];
+    perimeter += openstudio::getDistance(p1, p2);
+  }
+
+  return perimeter;
+}
+
+bool Polygon3d::isClockwise() {
+  if (OptionalVector3d normal = getOutwardNormal(m_outerPath)) {
+    return normal.get().z() > 0;
+  }
+
+  return true;
+}
+
+Point3d Polygon3d::centroid() {
+  if (auto p = openstudio::getCentroid(m_outerPath)) {
+    return p.get();
+  }
+
+  return Point3d();
+}
+
+bool Polygon3d::pointInPolygon(const Point3d& point, double tol) {
+  bool inside = false;
+  if (openstudio::pointInPolygon(point, m_outerPath, tol)) {
+    inside = true;
+  } else {
+    for (const auto& innerPath : m_innerPaths) {
+      if (openstudio::pointInPolygon(point, innerPath, tol)) {
+        inside = true;
+        break;
+      }
+    }
+  }
+  return inside;
+}
+
+bool Polygon3d::within(const Point3d& point, double tol) {
+  bool inside = false;
+  if (openstudio::within(point, m_outerPath, tol)) {
+    inside = true;
+    for (const auto& innerPath : m_innerPaths) {
+      if (openstudio::within(point, innerPath, tol)) {
+        inside = false;
+        break;
+      }
+    }
+  }
+  return inside;
+}
+
+bool Polygon3d::inside(const Point3d& point, double tol) {
+  return pointInPolygon(point, tol) || within(point, tol);
+}
+
+size_t Polygon3d::getSize() {
+  return m_outerPath.size();
+}
+
+Point3dVector Polygon3d::getEdge(size_t i) {
+  Point3d p1 = m_outerPath[i];
+  Point3d p2 = m_outerPath[(i + 1) % m_outerPath.size()];
+  return Point3dVector{p1, p2};
+}
+
+std::vector<Point3dVector> Polygon3d::overlap(const Point3dVector& line) const {
```

Comment:
This handles the case of a line intersecting a U shaped polygon twice?  Does it find intersections with holes too?  What is this used for?

Reply:
Yep twice or as many times as necessary

It finds all the segments on a polygon that overlap a line, so the polygon would be the building story footprint and the line would be an edge of a floor surface and the overlapping segments would give you the total length of the surface that is coincident with the building story shape, the exposed part of the floor.